纽约市Airbnb数据集

airbnb_NYC_2019.csv

简介：该数据集包含了2019年纽约市的Airbnb上线的房间情况。Airbnb是一个旅行房屋租赁社区，用户可通过网站或手机APP发布、搜索度假房屋租赁信息并在线预定。
变量含义：
- id：房间id
- name：房间名称
- host_id：房东id
- host_name：房东姓名
- neighbourhood_group：地区
- neighbourhood：街区
- latitude：纬度坐标
- longitude：经度坐标
- room_type：房间类型
- price：价格（美元）
- minimum_nights：最少预定夜晚数
- number_of_reviews：评论数量
- last_review：最新浏览时间
- reviews_per_month：每月浏览次数
- calculated_host_listings_count：房东挂出房子的数量
- availability_365：可预定房源的天数

目的：评估哪些街区的房源销售情况较好。

In [1]:
import pandas as pd

# 1 导入数据

In [2]:
o_data = pd.read_csv("./airbnb_NYC_2019.csv")
o_data

id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaN                NaN                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[48895 rows x 16 columns]

# 2 评估数据

## 2.1 评估数据整洁度

In [3]:
o_data.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [4]:
o_data.sample(10)

id                                             name    host_id  \
12034   9365025                  Light filled room in Greenpoint    2376644   
40866  31761090                 Wall-size window Private Bedroom  238145460   
46167  35125696  Cozy 1 bedroom w/twin bed & futon in Park Slope   14518163   
35532  28206177                          Bronx Brand New Room #3  203266238   
30870  23901420   Wyndham Midtown 45 (1 Bedroom Presidential) 2A  100238132   
42219  32782675                                  Clean and quite  140125433   
47743  35886159                         Ozone Park Tudor Village  269901856   
29592  22714571      Amazing Private Room @Madison Square Garden  167270001   
311       72265                  Private room in cozy Greenpoint     340692   
36438  28971081                   MOXY NYC DOWNTOWN-7 NIGHTS MIN    5144567   

          host_name neighbourhood_group   neighbourhood  latitude  longitude  \
12034     Katherine            Brooklyn      Greenpoint  40.72741  -73.94337   
40866         Cindy            Brooklyn        Flatbush  40.63037  -73.95671   
46167  Pedro & Jeff            Brooklyn     South Slope  40.66154  -73.98685   
35532         Steve               Bronx     Morris Park  40.85361  -73.85018   
30870       Michael           Manhattan         Midtown  40.75226  -73.97356   
42219          Yany               Bronx  Williamsbridge  40.87498  -73.85644   
47743           Akm              Queens      Ozone Park  40.67405  -73.85402   
29592         Piero           Manhattan  Hell's Kitchen  40.75590  -73.99324   
311         Vanessa            Brooklyn      Greenpoint  40.72646  -73.95341   
36438           Yun           Manhattan    Civic Center  40.71193  -74.00694   

             room_type  price  minimum_nights  number_of_reviews last_review  \
12034     Private room     80               2                  0         NaN   
40866     Private room     45               1                  9  2019-06-24   
46167     Private room     49               3                  2  2019-06-29   
35532     Private room     85               5                  5  2019-05-31   
30870  Entire home/apt    269               3                  5  2019-05-03   
42219     Private room     39               3                  6  2019-07-05   
47743     Private room     80              15                  0         NaN   
29592     Private room    109               1                 91  2019-06-25   
311       Private room     59               3                 29  2019-06-16   
36438     Private room    169               7                  4  2019-06-15   

       reviews_per_month  calculated_host_listings_count  availability_365  
12034                NaN                               1                 0  
40866               1.72                               1               129  
46167               1.62                               1                30  
35532               0.49                               5               365  
30870               0.36                              12                 0  
42219               1.57                               1                49  
47743                NaN                               1                61  
29592               5.11                               2               304  
311                 0.36                               1                15  
36438               1.35                              13               351

从显示出来的部分数据看数据符合：每一列只有一个变量、每一行只是哟个观测值、每一个单元格只有一个值。初步判断数据是整洁的。

# 2.2 评估数据的干净度

**初步观察**

In [5]:
o_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

观测值总共有48895个，变量`name`、`host_name`、`last_review`和`reviews_per_month`有缺失值，但是变量`name`和`host_name`的缺失值相比于总观测值数可以忽略，`last_review`和`reviews_per_month`缺失值较多并且他们的缺失值的数量还是相同的。

### 2.2.1 评估重复值

本数据的所有变量都可以不是唯一标示符，所以即使出现了重复也没有关系。

### 2.2.2 评估缺失值

有缺失值的变量是`name`、`host_name`、`last_review`和`reviews_per_month`

In [6]:
o_data[o_data["name"].isnull()]

id name   host_id host_name neighbourhood_group  \
2854    1615764  NaN   6676776     Peter           Manhattan   
3703    2232600  NaN  11395220      Anna           Manhattan   
5775    4209595  NaN  20700823     Jesse           Manhattan   
5975    4370230  NaN  22686810   Michaël           Manhattan   
6269    4581788  NaN  21600904     Lucie            Brooklyn   
6567    4756856  NaN   1832442  Carolina            Brooklyn   
6605    4774658  NaN  24625694      Josh           Manhattan   
8841    6782407  NaN  31147528  Huei-Yin            Brooklyn   
11963   9325951  NaN  33377685  Jonathan           Manhattan   
12824   9787590  NaN  50448556    Miguel           Manhattan   
13059   9885866  NaN  37306329  Juliette           Manhattan   
13401  10052289  NaN  49522403   Vanessa            Brooklyn   
15819  12797684  NaN  69715276       Yan           Manhattan   
16071  12988898  NaN  71552588    Andrea               Bronx   
18047  14135050  NaN  85288337      Jeff            Brooklyn   
28889  22275821  NaN  49662398  Kathleen            Brooklyn   

            neighbourhood  latitude  longitude        room_type  price  \
2854    Battery Park City  40.71239  -74.01620  Entire home/apt    400   
3703         East Village  40.73215  -73.98821  Entire home/apt    200   
5775    Greenwich Village  40.73473  -73.99244  Entire home/apt    225   
5975               Nolita  40.72046  -73.99550  Entire home/apt    215   
6269         Williamsburg  40.71370  -73.94378     Private room    150   
6567             Bushwick  40.70046  -73.92825     Private room     70   
6605   Washington Heights  40.85198  -73.93108     Private room     40   
8841         Williamsburg  40.71354  -73.93882     Private room     45   
11963      Hell's Kitchen  40.76436  -73.98573  Entire home/apt    190   
12824              Harlem  40.80316  -73.95189  Entire home/apt    300   
13059           Chinatown  40.71632  -73.99328     Private room     67   
13401         Brownsville  40.66409  -73.92314     Private room     50   
15819     Upper West Side  40.79843  -73.96404     Private room    100   
16071             Fordham  40.86032  -73.88493      Shared room    130   
18047  Bedford-Stuyvesant  40.69421  -73.93234     Private room     70   
28889            Bushwick  40.69546  -73.92741  Entire home/apt    110   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
2854             1000                  0         NaN                NaN   
3703                1                 28  2015-06-08               0.45   
5775                1                  1  2015-01-01               0.02   
5975                7                  5  2016-01-02               0.09   
6269                1                  0         NaN                NaN   
6567                1                  0         NaN                NaN   
6605                1                  0         NaN                NaN   
8841                1                  0         NaN                NaN   
11963               4                  1  2016-01-05               0.02   
12824               5                  0         NaN                NaN   
13059               4                  0         NaN                NaN   
13401               3                  3  2016-08-18               0.07   
15819               1                  0         NaN                NaN   
16071               1                  0         NaN                NaN   
18047               3                  0         NaN                NaN   
28889               4                  5  2018-08-13               0.27   

       calculated_host_listings_count  availability_365  
2854                                1               362  
3703                                1               341  
5775                                1                 0  
5975                                1                 0  
6269                                1                 0  
6567                                1    

`name`和`host_name`对于房源销售量的影响，不会是主要因素，因此即使缺失对之后的分析影响也不会有。

In [7]:
o_data[o_data["last_review"].isnull()]

id                                               name   host_id  \
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
19         7750                  Huge 2 BR Upper East  Cental Park     17985   
26         8700  Magnifique Suite au N de Manhattan - vue Cloitres     26394   
36        11452                        Clean and Quiet in Brooklyn      7355   
38        11943                          Country space in the city     45445   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

             host_name neighbourhood_group       neighbourhood  latitude  \
2            Elisabeth           Manhattan              Harlem  40.80902   
19                Sing           Manhattan         East Harlem  40.79685   
26     Claude & Sophie           Manhattan              Inwood  40.86754   
36                  Vt            Brooklyn  Bedford-Stuyvesant  40.68876   
38             Harriet            Brooklyn            Flatbush  40.63702   
...                ...                 ...                 ...       ...   
48890          Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891          Marisol            Brooklyn            Bushwick  40.70184   
48892    Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893              Taz           Manhattan      Hell's Kitchen  40.75751   
48894       Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
2      -73.94190     Private room    150               3                  0   
19     -73.94872  Entire home/apt    190               7                  0   
26     -73.92639     Private room     80               4                  0   
36     -73.94312     Private room     35              60                  0   
38     -73.96327     Private room    150               1                  0   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
2             NaN                NaN                               1   
19            NaN                NaN                               2   
26            NaN                NaN                               1   
36            NaN                NaN                               1   
38            NaN                NaN                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
2                   365  
19                  249  
26                    0  
36                  365  
38                  365  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[10052 rows 

`last_review`变量有10052条缺失的。

从上面显示的数据可以看出，`last_review`为NaN时`reviews_per_month`也为NaN并且`number_of_reviews`为0。我们猜想：
- (1)凡是`last_review`为NaN那么`reviews_per_month`也为NaN。
- (2)凡是`last_review`为NaN那么`reviews_per_month`也为NaN并且`number_of_reviews`为0。

接下来验证这个猜想。

In [8]:
(o_data["last_review"].isnull() != o_data["reviews_per_month"].isnull()).sum()

0

由上的出结论：凡是`last_review`为NaN那么`reviews_per_month`也为NaN。

In [9]:
o_data[o_data["last_review"].isnull() & (o_data["number_of_reviews"]!=0)]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

在`last_review`为缺失值的情况下`number_of_reviews`不为0的情况不存在，也就是说`last_review`为缺失值的情况下`number_of_reviews`一定为0。

猜想(2)也的证。

### 2.2.3 评估不一致数据

### 2.2.4 评估无效与错误数据

In [10]:
o_data.describe()

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

从统计描述来看没有不一致和错误数据。

# 3 清理数据

依据前面的评估需要做出如下清理：
1. `last_review`缺失值部分删除
2. `id`和`host_id`需要有整数转换为字符串
3. `last_review`需要转换为datetime数据类型

In [11]:
c_data = o_data.copy()
c_data

id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaN                NaN                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[48895 rows x 16 columns]

**1. `last_review`缺失值部分删除**

In [12]:
c_data.dropna(subset=["last_review"])
c_data = c_data.dropna(subset=["last_review"])
c_data

id                                              name    host_id  \
0          2539                Clean & quiet apt home by the park       2787   
1          2595                             Skylit Midtown Castle       2845   
3          3831                   Cozy Entire Floor of Brownstone       4869   
4          5022  Entire Apt: Spacious Studio/Loft by central park       7192   
5          5099         Large Cozy 1 BR Apartment In Midtown East       7322   
...         ...                                               ...        ...   
48782  36425863        Lovely Privet Bedroom with Privet Restroom   83554966   
48790  36427429                          No.2 with queen size bed  257683179   
48799  36438336                                   Seas The Moment  211644523   
48805  36442252                     1B-1B apartment near by Metro  273841667   
48852  36455809           Cozy Private Room in Bushwick, Brooklyn   74162901   

         host_name neighbourhood_group    neighbourhood  latitude  longitude  \
0             John            Brooklyn       Kensington  40.64749  -73.97237   
1         Jennifer           Manhattan          Midtown  40.75362  -73.98377   
3      LisaRoxanne            Brooklyn     Clinton Hill  40.68514  -73.95976   
4            Laura           Manhattan      East Harlem  40.79851  -73.94399   
5            Chris           Manhattan      Murray Hill  40.74767  -73.97500   
...            ...                 ...              ...       ...        ...   
48782        Rusaa           Manhattan  Upper East Side  40.78099  -73.95366   
48790         H Ai              Queens         Flushing  40.75104  -73.81459   
48799          Ben       Staten Island      Great Kills  40.54179  -74.14275   
48805       Blaine               Bronx       Mott Haven  40.80787  -73.92400   
48852    Christine            Brooklyn         Bushwick  40.69805  -73.92801   

             room_type  price  minimum_nights  number_of_reviews last_review  \
0         Private room    149               1                  9  2018-10-19   
1      Entire home/apt    225               1                 45  2019-05-21   
3      Entire home/apt     89               1                270  2019-07-05   
4      Entire home/apt     80              10                  9  2018-11-19   
5      Entire home/apt    200               3                 74  2019-06-22   
...                ...    ...             ...                ...         ...   
48782     Private room    129               1                  1  2019-07-07   
48790     Private room     45               1                  1  2019-07-07   
48799     Private room    235               1                  1  2019-07-07   
48805  Entire home/apt    100               1                  2  2019-07-07   
48852     Private room     30               1                  1  2019-07-08   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
3                   4.64                               1               194  
4                   0.10                               1                 0  
5                   0.59                               1               129  
...                  ...                             ...               ...  
48782               1.00                               1               147  
48790               1.00                               6               339  
48799               1.00                               1                87  
48805               2.00                               1                40  
48852               1.00                               1                 1  

[38843 rows x 16 columns]

**2. `id`和`host_id`需要有整数转换为字符串**

In [13]:
c_data["id"].astype(str)

0            2539
1            2595
3            3831
4            5022
5            5099
           ...   
48782    36425863
48790    36427429
48799    36438336
48805    36442252
48852    36455809
Name: id, Length: 38843, dtype: object

In [14]:
c_data["id"] = c_data["id"].astype(str)
c_data["id"]

/var/folders/4c/0q2w7hgn3rn1kfmhshws6pjm0000gn/T/ipykernel_68111/3078278774.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  c_data["id"] = c_data["id"].astype(str)


0            2539
1            2595
3            3831
4            5022
5            5099
           ...   
48782    36425863
48790    36427429
48799    36438336
48805    36442252
48852    36455809
Name: id, Length: 38843, dtype: object

In [15]:
c_data["host_id"] = c_data["host_id"].astype(str)

/var/folders/4c/0q2w7hgn3rn1kfmhshws6pjm0000gn/T/ipykernel_68111/372837701.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  c_data["host_id"] = c_data["host_id"].astype(str)


**3. `last_review`需要转换为datetime数据类型**

In [16]:
pd.to_datetime(c_data["last_review"])

0       2018-10-19
1       2019-05-21
3       2019-07-05
4       2018-11-19
5       2019-06-22
           ...    
48782   2019-07-07
48790   2019-07-07
48799   2019-07-07
48805   2019-07-07
48852   2019-07-08
Name: last_review, Length: 38843, dtype: datetime64[ns]

In [17]:
c_data["last_review"] = pd.to_datetime(c_data["last_review"])
c_data["last_review"]

/var/folders/4c/0q2w7hgn3rn1kfmhshws6pjm0000gn/T/ipykernel_68111/377306588.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  c_data["last_review"] = pd.to_datetime(c_data["last_review"])


0       2018-10-19
1       2019-05-21
3       2019-07-05
4       2018-11-19
5       2019-06-22
           ...    
48782   2019-07-07
48790   2019-07-07
48799   2019-07-07
48805   2019-07-07
48852   2019-07-08
Name: last_review, Length: 38843, dtype: datetime64[ns]

In [18]:
c_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38843 entries, 0 to 48852
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              38843 non-null  object        
 1   name                            38837 non-null  object        
 2   host_id                         38843 non-null  object        
 3   host_name                       38827 non-null  object        
 4   neighbourhood_group             38843 non-null  object        
 5   neighbourhood                   38843 non-null  object        
 6   latitude                        38843 non-null  float64       
 7   longitude                       38843 non-null  float64       
 8   room_type                       38843 non-null  object        
 9   price                           38843 non-null  int64         
 10  minimum_nights                  38843 non-null  int64         
 11  nu

# 4 保存数据

In [19]:
c_data.to_csv("cleaded_airbnb.csv", index=False)

In [20]:
pd.read_csv("cleaded_airbnb.csv")

id                                              name    host_id  \
0          2539                Clean & quiet apt home by the park       2787   
1          2595                             Skylit Midtown Castle       2845   
2          3831                   Cozy Entire Floor of Brownstone       4869   
3          5022  Entire Apt: Spacious Studio/Loft by central park       7192   
4          5099         Large Cozy 1 BR Apartment In Midtown East       7322   
...         ...                                               ...        ...   
38838  36425863        Lovely Privet Bedroom with Privet Restroom   83554966   
38839  36427429                          No.2 with queen size bed  257683179   
38840  36438336                                   Seas The Moment  211644523   
38841  36442252                     1B-1B apartment near by Metro  273841667   
38842  36455809           Cozy Private Room in Bushwick, Brooklyn   74162901   

         host_name neighbourhood_group    neighbourhood  latitude  longitude  \
0             John            Brooklyn       Kensington  40.64749  -73.97237   
1         Jennifer           Manhattan          Midtown  40.75362  -73.98377   
2      LisaRoxanne            Brooklyn     Clinton Hill  40.68514  -73.95976   
3            Laura           Manhattan      East Harlem  40.79851  -73.94399   
4            Chris           Manhattan      Murray Hill  40.74767  -73.97500   
...            ...                 ...              ...       ...        ...   
38838        Rusaa           Manhattan  Upper East Side  40.78099  -73.95366   
38839         H Ai              Queens         Flushing  40.75104  -73.81459   
38840          Ben       Staten Island      Great Kills  40.54179  -74.14275   
38841       Blaine               Bronx       Mott Haven  40.80787  -73.92400   
38842    Christine            Brooklyn         Bushwick  40.69805  -73.92801   

             room_type  price  minimum_nights  number_of_reviews last_review  \
0         Private room    149               1                  9  2018-10-19   
1      Entire home/apt    225               1                 45  2019-05-21   
2      Entire home/apt     89               1                270  2019-07-05   
3      Entire home/apt     80              10                  9  2018-11-19   
4      Entire home/apt    200               3                 74  2019-06-22   
...                ...    ...             ...                ...         ...   
38838     Private room    129               1                  1  2019-07-07   
38839     Private room     45               1                  1  2019-07-07   
38840     Private room    235               1                  1  2019-07-07   
38841  Entire home/apt    100               1                  2  2019-07-07   
38842     Private room     30               1                  1  2019-07-08   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                   4.64                               1               194  
3                   0.10                               1                 0  
4                   0.59                               1               129  
...                  ...                             ...               ...  
38838               1.00                               1               147  
38839               1.00                               6               339  
38840               1.00                               1                87  
38841               2.00                               1                40  
38842               1.00                               1                 1  

[38843 rows x 16 columns]

In [21]:
pd.read_csv("cleaded_airbnb.csv").info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38843 entries, 0 to 38842
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              38843 non-null  int64  
 1   name                            38837 non-null  object 
 2   host_id                         38843 non-null  int64  
 3   host_name                       38827 non-null  object 
 4   neighbourhood_group             38843 non-null  object 
 5   neighbourhood                   38843 non-null  object 
 6   latitude                        38843 non-null  float64
 7   longitude                       38843 non-null  float64
 8   room_type                       38843 non-null  object 
 9   price                           38843 non-null  int64  
 10  minimum_nights                  38843 non-null  int64  
 11  number_of_reviews               38843 non-null  int64  
 12  last_review                     

重新导入数据这里`last_review`的数据类型仍然不是datetime